contrib/OtlpGrpc/Exporter.php

```diff
@@ -0,0 +1,285 @@
+                'instrumentation_library' => $il,
+                'spans' => [$convertedSpan]
+            ]);
+
+            array_push($ilspans, $ilspan);
+        }
+
+
+        $resourcespans = new Proto\Trace\V1\ResourceSpans([
+            'instrumentation_library_spans' => $ilspans
+        ]);
+
+        
+        $request= new V1\ExportTraceServiceRequest();
+        $request->setResourceSpans([$resourcespans]);
+
+
+        list($response, $status) = $this->client->Export($request)->wait();
+        if ($status->code !== Grpc\STATUS_OK) {
+            echo "ERROR: " . $status->code . ", " . $status->details . PHP_EOL;
+        }
+
+
+        // TODO: Make this work
+        //echo $response->getMessage() . PHP_EOL;
+
+        // if ($response->getStatusCode() >= 400 && $response->getStatusCode() < 500) {
+        //     return Trace\Exporter::FAILED_NOT_RETRYABLE;
+        // }
+
+        // if ($response->getStatusCode() >= 500 && $response->getStatusCode() < 600) {
+        //     return Trace\Exporter::FAILED_RETRYABLE;
+        // }
+
+        return Trace\Exporter::SUCCESS;
+    }
+
+    public function as_otlp_key_value($key, $value): KeyValue {
+        return new KeyValue([
+            'key' => $key,
+            'value' => $this->as_otlp_any_value($value)
+        ]);
+    }
+
+    public function as_otlp_any_value($value): AnyValue
+    {
+        $result = new AnyValue();
+
+        switch (true) {
+            case is_array($value):
+                $result->setArrayValue($value);
+                break;
+            case is_int($value):
+                $result->setIntValue($value);
+                break;
+            case is_bool($value):
+                $result->setBoolValue($value);
+                break;
+            case is_double($value):
+                $result->setDoubleValue($value);
+                break;
+            case is_string($value):
+                $result->setStringValue($value);
+                break;
+        }
+
+        return $result;
+    }
+
+    public function as_otlp_span_kind($kind): int
+    {
+
+        switch ($kind) {
+            case 0: return SpanKind::SPAN_KIND_INTERNAL;
+            case 1: return SpanKind::SPAN_KIND_CLIENT;
+            case 2: return SpanKind::SPAN_KIND_SERVER;
+            case 3: return SpanKind::SPAN_KIND_PRODUCER;
+            case 4: return SpanKind::SPAN_KIND_CONSUMER;
+        }
+
+        return SpanKind::SPAN_KIND_UNSPECIFIED;
+    }
+
+    // ["trace_id":protected]=> string(32) "f8ebf54644f65fae5e98c5dd84b9d196"
+    // ["span_id":protected]=> string(16) "b90a8323cfd7cd66"
+    public function as_otlp_span(Span $span): CollectorSpan
+    {
+
+        $duration_ns = (($span->getEnd() - $span->getStart()));
+        $end_timestamp = ($span->getStartEpochTimestamp() + $duration_ns);
+
+        $row = [
+            // TODO: Fix this, I was getting 32, and 16 byte ids sent to the collector which it didn't like
```

Comment:
I think this is part of the spec?  What problems were you facing here?

Reply:
According to the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md), the collector expects the following format:
```
JSON-encoded Protobuf payloads use proto3 standard defined 
JSON Mapping for mapping between Protobuf and JSON, with one deviation 
from that mapping: the trace_id and span_id byte arrays are represented as 
case-insensitive hex-encoded strings, they are not base64-encoded like it is 
defined in the standard JSON Mapping. The hex encoding is used for trace_id 
and span_id fields in all OTLP Protobuf messages, e.g. 
the Span, Link, LogRecord, etc. messages.
```
You can see Ruby handled this in the [background](https://github.com/open-telemetry/opentelemetry-ruby/blob/73952a92f5117cab3c6a8264406297602102d628/sdk/test/opentelemetry/sdk/trace/samplers_test.rb#L204).

Whereas python handled this more [explicitly](https://github.com/open-telemetry/opentelemetry-python/blob/main/exporter/opentelemetry-exporter-otlp/src/opentelemetry/exporter/otlp/trace_exporter/__init__.py#L137).

Reply:
Thank you both, I assumed there was something more going on. The links are useful